geometric_kernels/spaces/special_orthogonal_group.py

```diff
@@ -0,0 +1,408 @@
+                    for k, pk in enumerate(signature)
+                ]
+                rep_dim = int(
+                    reduce(
+                        operator.mul,
+                        (2 / math.factorial(2 * k) for k in range(1, self.rank)),
+                    )
+                    * reduce(
+                        operator.mul,
+                        (
+                            (qs[i] - qs[j]) * (qs[i] + qs[j])
+                            for i, j in it.combinations(range(self.rank), 2)
+                        ),
+                        1,
+                    )
+                )
+                return int(round(rep_dim))
+
+        def _compute_eigenvalue(sgn):
+            np_sgn = np.array(sgn)
+            return (
+                np.linalg.norm(self.rho + np_sgn) ** 2 - np.linalg.norm(self.rho) ** 2
+            )
+
+        signatures_vals = []
+        for sgn in signatures:
+            dim = _compute_dim(sgn)
+            eigenvalue = _compute_eigenvalue(sgn)
+            signatures_vals.append([sgn, dim, eigenvalue])
+
+        signatures_vals.sort(key=lambda x: x[2])  # sort by eigenvalue
+        signatures_vals = signatures_vals[:num_repr]
+
+        signatures = np.array([x[0] for x in signatures_vals])  # [M, R]
+        dims = np.array([x[1] for x in signatures_vals])  # [M, ]
+        eigenvalues = np.array([x[2] for x in signatures_vals])  # [M, ]
+
+        return signatures, dims, eigenvalues
+
+
+class SpecialOrthogonalGroup(
+    DiscreteSpectrumSpace, gs.geometry.special_orthogonal._SpecialOrthogonalMatrices
+):
+    r"""
+    Special orthogonal group `SO(n)`, that is a group of orthogonal matrices
+    with determinant equal to 1.
+    """
+
+    def __init__(self, n: int):
+        r"""
+        :param n: Dimensionality `SO(n)`, aka the size of the matrices.
+        """
+        self.cache: Dict[int, Any] = {}
+        super().__init__(n=n)
+
+    @property
+    def dimension(self):
+        return self.n
+
+    def get_eigenvalues(self, num: int):
+        """
+        First `num` eigenvalues.
+
+        :param num: number of eigenvalues returned.
+
+        :return: [num, 1] array containing the eigenvalues.
+        """
+        if num not in self.cache:
+            eigenfunctions = SOEigenfunctions(self.n, num)
+            eigenvalues = eigenfunctions.repr_eigenvalues ** 2
+            self.cache[num] = eigenfunctions
+            return B.reshape(eigenvalues, -1, 1)
+        else:
+            eigenfunctions = self.cache[num]
+            eigenvalues = eigenfunctions.repr_eigenvalues ** 2
+            return B.reshape(eigenvalues, -1, 1)
+
+    def get_eigenfunctions(self, num):
+        """
+        :param num: number of eigenfunctions returned.
+        """
+        if num not in self.cache:
+            eigenfunctions = SOEigenfunctions(self.dim, num)
+            self.cache[num] = eigenfunctions
+            return eigenfunctions
+        else:
+            return self.cache[num]
```

Comment:
can we use lru cache for this: https://docs.python.org/3/library/functools.html?